compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -274,77 +274,139 @@ void deadCodeElimination(FnSymbol* fn) {
+*                                                                             *
+* The hardest part is determining when this code needs to be revisited.       *
+*                                                                             *
+************************************** | *************************************/
+
+static bool isDeadStringLiteral(VarSymbol* string);
+static void removeDeadStringLiteral(DefExpr* defExpr);
+
 static void deadStringLiteralElimination() {
+  int numStmt        = 0;
+  int numDeadLiteral = 0;
 
-  // find all the symExprs in the string literal module
-  std::vector<SymExpr*> symExprs;
-  collectSymExprs(stringLiteralModule, symExprs);
-
-  for_vector(SymExpr, stringUse, symExprs) {
-    Symbol* stringSym = stringUse->symbol();
-    // if we're looking at a Chapel string created from a string literal
-    if (stringSym->hasFlag(FLAG_CHAPEL_STRING_LITERAL)) {
-      // and there's only a single use of it
-      int nuses = stringSym->countUses(/*max*/ 2);
-      if (nuses == 1) {
-        // then that use is the RHS of `ret_to_arg_ref_tmp = &str_literal_id`,
-        // which is only used in the string constructor so the string is dead. 
-        for_SymbolUses(use, stringSym) {
-          // this loop should only run once
-          assert(use == stringUse);
-        }
+  for_alist(stmt, stringLiteralModule->block->body) {
+    numStmt = numStmt + 1;
 
-        // gather the AST used to create a Chapel string from the literal,
-        // using variable names that mimic the current generated code
-        CallExpr*      ret_to_arg_assign = toCallExpr(stringUse->getStmtExpr());
-        SymExpr*       ret_to_arg        = toSymExpr(ret_to_arg_assign->get(1));
-        SymExpr*       ret_to_arg_use    = ret_to_arg->symbol()->getSingleUse();
-        INT_ASSERT(ret_to_arg_use != NULL);
-        CallExpr*      stringCtor        = toCallExpr(ret_to_arg_use->parentExpr);
-        SymExpr*       call_tmp          = toSymExpr(stringCtor->get(1));
-        SymExpr*       call_tmp_def      = call_tmp->symbol()->getSingleDef();
-        INT_ASSERT(call_tmp_def != NULL);
-        CallExpr*      call_tmp_assign   = toCallExpr(call_tmp_def->parentExpr);
-
-        // remove all the AST, in the order listed in the function comment
-        stringUse->symbol()->defPoint->remove();
-        call_tmp->symbol()->defPoint->remove();
-        call_tmp_assign->remove();
-        ret_to_arg->remove();
-        ret_to_arg_assign->remove();
-        stringCtor->remove();
+    if (DefExpr* defExpr = toDefExpr(stmt)) {
+      if (VarSymbol* symbol = toVarSymbol(defExpr->sym)) {
+        if (isDeadStringLiteral(symbol) == true) {
+          removeDeadStringLiteral(defExpr);
+
+          numDeadLiteral = numDeadLiteral + 1;
+        }
       }
     }
   }
+
+  // Confirm that the analysis found some dead literals
+  // Ignore this test for --minimal-modules
+  if (numStmt > 500) {
+    INT_ASSERT((1.0 * numDeadLiteral) / numStmt > 0.2);
```

Comment:
Instead of having the check in the compiler like this, what do you think about having a `--report-dead-string-literals` flag (like `--report-dead-modules`) and just creating a test like optimizations/deadCodeElimination/elliot/countDeadModules. The number of dead string literals will probably change pretty frequently, so it should probably have some sort of threshold.

Alternatively not eliminating dead strings really blows up the generated code size so we could just rely on the generated code size compiler graph to catch this after the fact. 

And also, sorry. This is definetly not code that I'm proud of having written.